hy/macros.py

```diff
@@ -8,30 +8,75 @@
 import pkgutil
 import traceback
 
+from funcparserlib.parser import NoParseError
+
 from hy._compat import PY3_8
-from hy.models import replace_hy_obj, Expression, Symbol, as_model
+from hy.model_patterns import whole
+from hy.models import replace_hy_obj, Expression, Symbol, as_model, is_unpack
 from hy.lex import mangle, unmangle
 from hy.errors import (HyLanguageError, HyMacroExpansionError, HyTypeError,
                        HyRequireError)
 
-EXTRA_MACROS = [
-    "hy.core.macros",
-]
+EXTRA_MACROS = ["hy.core.result_macros", "hy.core.macros"]
 
 
 def macro(name):
     """Decorator to define a macro called `name`.
     """
-    name = mangle(name)
-    def _(fn):
-        fn = rename_function(fn, name)
+    return lambda fn: install_macro(name, fn, fn)
+
+
+def pattern_macro(names, pattern, shadow = None):
+    pattern = whole(pattern)
+    py_version_required = None
+    if isinstance(names, tuple):
+        py_version_required, names = names
+
+    def dec(fn):
+
+        def wrapper_maker(name):
+            def wrapper(hy_compiler, *args):
+
+                if (shadow and
+                        any(is_unpack("iterable", x) for x in args)):
+                    # Try a shadow function call with this name instead.
+                    return Expression([
+                        Symbol('hy.core.shadow.' + name),
+                        *args]).replace(hy_compiler.this)
 
-        module = inspect.getmodule(fn)
-        module_macros = module.__dict__.setdefault('__macros__', {})
-        module_macros[name] = fn
+                expr = hy_compiler.this
+                root = unmangle(expr[0])
```

Comment:
The macros need to set the source-line positions when they're constructing the python AST objects. By the time they've returned an AST object up the call stack, it's too late for subordinate AST objects to have their positions set correctly.

Reply:
Would `ast.fix_missing_locations()` solve that problem?

Reply:
Hmm, I think so. I would try an `ast.fix_missing_locations()` step after each macro-expansion that produces a `Result` or `AST`, with a temporary outermost AST object that has the location information to be copied.

My one concern is that once you can write `ast.X(…)`, with no location information, in place of `asty.X(expr, …)`, people working on Hy's core macros will be liable to mistakenly write `ast.X(…)` when they meant `asty.X(some_model, …)`. There won't be a crash; nobody will notice anything amiss unless they see an error message with a position that's not as far into the parse tree as it should be. Do you think getting rid of the `expr` parameter and replacing a lot of `asty.X(expr, …)` with `ast.X(…)` is worth these potential bugs?

Reply:
If the compiler calls sets the position on the result of `macroexpand()` from the original `expr`, then calls `ast.fix_missing_locations()`, then none of the core macros will even need to use `asty` at all, unless I'm missing something.

Reply:
You'll still need to set locations explicitly in order to point to arguments of the macro, or subcomponents of the arguments, instead of the whole form.

Reply:
Ah yes, that's a good point.
Hmm, I still would prefer a way of not needing to set extra state in the compiler, but I'll leave that for a later PR.